CONTRIBUTING.md

```diff
@@ -0,0 +1,247 @@
+> subject of an email and the rest of the text as the body. The blank line
+> separating the summary from the body is critical (unless you omit the
+> body entirely); tools like rebase can get confused if you run the two
+> together.
+>
+> Write your commit message in the imperative: “Fix bug” and not “Fixed
+> bug” or “Fixes bug.” This convention matches up with commit messages
+> generated by commands like git merge and git revert.
+>
+> Further paragraphs come after blank lines.
+>
+> * Bullet points are okay, too
+>
+> * Typically a hyphen or asterisk is used for the bullet, followed by a
+>   single space, with blank lines in between, but conventions vary here
+>
+> * Use a hanging indent
+
+You should also feel free to use Markdown in the commit messages, as our project
+is hosted on GitHub which renders it (and Markdown is human readable).
+
+## Design Patterns
+
+The most important goal we are attempting to accomplish with LISAv3 is for it to
+be “simple, clean, and with a low maintenance cost.”
+
+We should use caution when using Object Oriented Design, because when it is used
+without critical analysis, it creates unmaintainable code. A great talk on this
+subject is [Stop Writing Classes](https://www.youtube.com/watch?v=o9pEzgHorH0),
+by Jack Diederich. As he says, “classes are great but they are also overused.”
+
+This [Python Design Patterns](https://python-patterns.guide/) is a fantastic
+collection of material for writing maintainable Python code. It specifically
+details many of the common “Object Oriented” patterns from the Gang of Four book
+(which, in fact, were patterns geared toward languages like C++, and no longer
+apply to modern languages like Python), what lessons can be learned from them,
+and how to apply them (or their modern alternatives) today. It also serves as an
+easy-to-read guide to the Gang of Four book itself, as its principles still
+serve us well today.
+
+Every time a developer chooses to use a design pattern, that person needs to
+reason through and document why it was chosen, and what alternatives were
+considered. We will recreate the problems with LISAv2 unless we take our time to
+carefully create a well-designed and maintainable framework.
+
+Several popular patterns that actually _do not_ work well in Python are:
+
+* [The Abstract Factory Pattern](https://python-patterns.guide/gang-of-four/abstract-factory/)
+* [The Factory Method Pattern](https://python-patterns.guide/gang-of-four/factory-method/)
+* [The Prototype Pattern](https://python-patterns.guide/gang-of-four/prototype/)
+* [The Singleton Pattern](https://python-patterns.guide/gang-of-four/singleton/)
+
+Conversely, patterns that are a natural fit to Python include:
+
+* [The Composite Pattern](https://python-patterns.guide/gang-of-four/composite/)
+* [The Flyweight Pattern](https://python-patterns.guide/gang-of-four/flyweight/)
+* [The Iterator Pattern](https://python-patterns.guide/gang-of-four/iterator/)
```

Comment:
I'd even move this one to the previous list, because although you can create iterators like the article mentions, you often just want to create generator functions.  `yield` and `yield from` are amazingly easy to use.  The `OddIterators` and `OddNumbers` class pair for instance is this tiny function:

```python
def odd_numbers(maximum=None):
    current = 1
    while True:
        if maximum is not None and current >= maximum:
             raise StopIteration
        yield current
        current += 2
```

Reply:
Indeed, generators are definitely the more powerful way to do this. Thanks @lpereira!